### ISSUE TITLE
Support eslint flat config

### ISSUE THREAD

**Issue by silverwind** [un-ts/eslint-plugin-import-x#29]
Likely depends on https://github.com/import-js/eslint-plugin-import/issues/2556 but it's good to track nonetheless.

Ref: https://github.com/eslint/eslint/issues/18093

**Comment by JounQin** [comment]
`eslint-plugin-i` is now `eslint-plugin-import-x`.

I'm focusing on TypeScript migration first, see also #41.

After that, I'll start working on new features like supporting flat config.

**Comment by meteorlxy** [comment]
Some rules will break when using flat config. See [this comment](https://github.com/eslint/eslint/issues/16485#issuecomment-1303446968)

In brief, eslint will fail to use `eslint.config.js` as config file after enabling `import/no-unused-modules` with `unusedExports: true`

Related issues:
- https://github.com/import-js/eslint-plugin-import/issues/2907
- https://github.com/import-js/eslint-plugin-import/issues/2964

**Comment by karlhorky** [comment]
Interesting, we are using `eslint-plugin-import-x` now with the Flat Config:

- https://github.com/import-js/eslint-plugin-import/issues/2948#issuecomment-2119214365

This seems to be working for us so far:

- [upleveled/eslint-config-upleveled@`3115d7e` (#369)](https://github.com/upleveled/eslint-config-upleveled/pull/369/commits/3115d7e081ce25a1a9821d1500f00acb66fca8fd)

Simplified version of the code:

```ts
import eslintImportX from 'eslint-plugin-import-x';

/** @type {import('@typescript-eslint/utils/ts-eslint').FlatConfig.ConfigArray} */
const configArray = [
  {
    plugins: {
     'import-x': eslintImportX,
    },
    settings: {
      'import-x/parsers': {
        '@typescript-eslint/parser': ['.ts', '.tsx'],
      },
      'import-x/resolver': {
        // Load <rootdir>/tsconfig.json
        typescript: true,
        node: true,
      },
    },
    rules: {
      // Error on imports that don't match the underlying file system
      // https://github.com/un-ts/eslint-plugin-import-x/blob/master/docs/rules/no-unresolved.md
      'import-x/no-unresolved': 'error',
  },
];
```

**Comment by Zamiell** [comment]
@karlhorky I am trying out this plugin with the flat config. I copy pasted the config from your post, but I get these errors:

![image](https://github.com/un-ts/eslint-plugin-import-x/assets/5511220/d2a42746-4fa7-4116-ace2-7a0486255e5b)

Any idea why?

**Comment by n0099** [comment]
```js
module.exports = {
    extends: [
        'plugin:import-x/recommended',
        'plugin:import-x/typescript',
    ],
}
```
can only be translate into
```js
export default [
    ...compat.config(pluginImportX.configs.recommended), // https://github.com/un-ts/eslint-plugin-import-x/issues/29
    ...compat.config(pluginImportX.configs.typescript),
];
```
since
https://github.com/un-ts/eslint-plugin-import-x/blob/3abe5e49683e0f973232bb631814b935e1ca7091/src/config/recommended.ts#L7
is still passing string as plugin.

**Comment by onx2** [comment]
@n0099 where is `compat` coming from?

**Comment by n0099** [comment]
> @n0099 where is `compat` coming from?

[`@eslint/eslintrc`](https://www.npmjs.com/package/@eslint/eslintrc) https://github.com/ota-meshi/typescript-eslint-parser-for-extra-files/issues/95#issuecomment-2148604881
```js
import { FlatCompat } from '@eslint/eslintrc';
```

> My super simple set up works for everything but the import sorting.

If you only want to extends from `typescript` rules, the plugin is not defined in the rule: https://github.com/un-ts/eslint-plugin-import-x/blob/3abe5e49683e0f973232bb631814b935e1ca7091/src/config/typescript.ts#L16-L35
so manually add plugin define to let eslint know where to find defines for rule with prefix `import-x/`
```diff
+  { plugins: { 'import-x': pluginImportX } },
  pluginImportX.configs.typescript,
```
or just also extending `recommended` rules:
```diff
+  ...compat.config(pluginImportX.configs.recommended),
  pluginImportX.configs.typescript,
```

**Comment by muuvmuuv** [comment]
Where do you import compat from?

**Comment by GalindoSVQ** [comment]
> Where do you import compat from?

https://github.com/un-ts/eslint-plugin-import-x/issues/29#issuecomment-2156223337

**Comment by KristjanTammekivi** [comment]
> Just to mention, some rules will break when using flat config. See [this comment](https://github.com/eslint/eslint/issues/16485#issuecomment-1303446968)
> 
> In brief, eslint will fail to use `eslint.config.js` as config file after enabling `import/no-unused-modules` with `unusedExports: true`
> 
> Related issues:
> 
>     * [Bug: can not find flat file config eslint/eslint#16485](https://github.com/eslint/eslint/issues/16485)
> 
>     * [import/no-unused-modules breaks `--no-eslintrc` and eslint fails import-js/eslint-plugin-import#2907](https://github.com/import-js/eslint-plugin-import/issues/2907)
> 
>     * [Add note(s) to docs about flat configs and/or no-unused-modules' `unusedExports` not being supported in flat configs import-js/eslint-plugin-import#2964](https://github.com/import-js/eslint-plugin-import/issues/2964)

Also this https://github.com/eslint/eslint/issues/18087

**Comment by SukkaW** [comment]
Let's deprecate `no-unused-modles` then~!

**Comment by SukkaW** [comment]
@Zamiell The discussion about deprecating `no-unused-modules` rule can be found here: https://github.com/un-ts/eslint-plugin-import-x/issues/90

**Comment by silverwind** [comment]
Never heard about knip, but I'm generally in favor of at least disabling rules that are duplicate with what typescript does, if the linted file is typescript, of course.

**Comment by GerroDen** [comment]
It works great in flat config 👍🏻  but `no-unused-modules` leads to an eslint error that the eslint config is missing.

**Comment by SukkaW** [comment]
> It works great in flat config 👍🏻 but `no-unused-modules` leads to an eslint error that the eslint config is missing.

Yes. This is because `no-unused-modules` uses an ESLint internal private API that was designed to be used for crawling eslintrc files.

**Comment by silverwind** [comment]
In https://github.com/import-js/eslint-plugin-import/pull/3018 they used `@nodelib/fs.walk` if I understand correctly, seems like a popular enough dependency to use.

**Comment by silverwind** [comment]
Maybe @michaelfaith would be interested in porting their work here in case that PR stalls because of those "node 4 concerns" which we do not have with this fork 😆.

**Comment by SukkaW** [comment]
> Maybe @michaelfaith would be interested in porting their work here in case that PR stalls because of those "node 4 concerns" which we do not have with this fork 😆.

Yeah. @michaelfaith if you are interested, you can port that PR to `eslint-plugin-import-x` as well. We use TypeScript and we are not limited to Node.js 4.

**Comment by michaelfaith** [comment]
Sure, I don't mind helping out, when I have some time.

**Comment by controversial** [comment]
> In https://github.com/import-js/eslint-plugin-import/pull/3018 they used `@nodelib/fs.walk` to replace it

note the `fs.walk` approach discussed in https://github.com/import-js/eslint-plugin-import/pull/3018 for replacing FileEnumerator depends on a [proposed eslint API](https://github.com/eslint/eslint/issues/18087) that’s currently only hypothetical

**Comment by SukkaW** [comment]
> that’s currently only hypothetical

Did you mean `context.session`?

> I’d support deprecating `no-unused-modules` and skipping support for using that rule with flat config

I've considered deprecating `no-unused-modules` (https://github.com/un-ts/eslint-plugin-import-x/issues/90) but I changed my idea.

And as long as we don't include this rule in the flat config, everything would be fine.

**Comment by silverwind** [comment]
I'd definitely keep it at this point, it's too valuable and not everyone wants to install another dependency and fight its config just to replace one rule.

**Comment by michaelfaith** [comment]
Yeah

> > that’s currently only hypothetical
> 
> Did you mean `context.session`?

Yeah, I wouldn't really say "hypothetical".  Proposed is accurate though.  The two functions that I used in that implementation were added in a POC branch of the eslint repo to prove the proposed api.  The idea is that the code would use those functions if they were on the context, and if not, fall back to the old implementation.  And if, between now and then, those functions get introduced with a slightly different API, then the implementation would need to be adjusted.  It'd be the same here, if that's the route you want to go.

**Comment by michaelfaith** [comment]
Oh, and to add a little more context: that api very likely would have already been added to ESLint in v9, if it weren't for the feet-dragging from the `eslint-plugin-import` project in validating the approach.  If the API is ever going to be added, there needs to be an expressed demand for it.  So if you want it, this could be a good way to express that demand.  Food for thought.

**Comment by michaelfaith** [comment]
I started a draft PR, with everything except the changes to `no-unused-module`.  I wanted to confirm that that's the direction you want to go before implementing it.
